Periodic Pinging and active fixLowPeers

aarshkshah1992 opened the pull request:
@aschmahmann 

For #521  and makes fixLowPeers more active.

RT peer at https://github.com/libp2p/go-libp2p-kbucket/pull/69.

Stebalien commented on subscriber_notifee.go:
```diff
@@ -125,7 +125,10 @@ func handlePeerIdentificationCompletedEvent(dht *IpfsDHT, e event.EvtPeerIdentif
 	} else if valid {
 		dht.peerFound(dht.ctx, e.Peer, false)
-		dht.fixLowPeers()
+		select {
```
nit: can we put this in a function?

aarshkshah1992 replied:
Done.

Stebalien commented on dht.go:
```diff
@@ -181,6 +188,10 @@ func New(ctx context.Context, h host.Host, options ...Option) (*IpfsDHT, error)
 	go dht.persistRTPeersInPeerStore()
+
+	// listens to the fix low peers chan and tries to fix the Routing Table
+	go dht.fixLowPeersRoutine()
```
later: given that we're using goprocess, let's manage this with goprocess.

aarshkshah1992 replied:
Done.

Stebalien commented on dht_bootstrap.go:
```diff
@@ -119,6 +120,20 @@ func (dht *IpfsDHT) startRefreshing() error {
+			for _, ps := range dht.routingTable.GetPeerInfos() {
+				// ping the peer if it's due for a ping and evict it if the ping fails
+				if time.Since(ps.LastSuccessfulOutboundQuery) > dht.maxLastSuccessfulOutboundThreshold {
+					livelinessCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
```
Can we make this a constant?

aarshkshah1992 replied:
Done.

Stebalien commented on dht_bootstrap.go:
```diff
@@ -119,6 +120,20 @@ func (dht *IpfsDHT) startRefreshing() error {
+			// ping Routing Table peers that haven't been hear of/from in the interval they should have been.
+			for _, ps := range dht.routingTable.GetPeerInfos() {
+				// ping the peer if it's due for a ping and evict it if the ping fails
+				if time.Since(ps.LastSuccessfulOutboundQuery) > dht.maxLastSuccessfulOutboundThreshold {
```
We're not treating this as a successful query. I think we need to record a _separate_ "last seen"? Otherwise, we'll keep re-connecting once the peer hits this limit.

aarshkshah1992 replied:
As discussed offline, this is not needed for now as we will probably evict flaky-connectivity peers from the DHT because they wont respond back to us with good latencies. Please see https://github.com/libp2p/go-libp2p-kad-dht/pull/536 for more details.

aarshkshah1992 commented:
@Stebalien I have addressed all the review comments and am merging this because of the approval. Lmk of concerns crop up and I'll address them in a new PR.
